### DISCUSSION TITLE
New inbound FBA API 2024-03-20 - How to match Amazon provided boxes labels to our boxes?  NOTE: Need API ability to provide SELLER BOX ID to print on labels

### DISCUSSION THREAD

**Discussion by MetaMan2023** [amzn/selling-partner-api-models#4365] [Q&A]
Summary of the question:

If we use the API to supply packing information (ie: setPackingInformation) to Amazon for an inbound plan / shipment -- how do we match the resulting box labels with actual boxes in our warehouse?   There is no way for us to provide an internal (seller) box identifier to Amazon.

Detailed example:

I understand now that boxes cannot be “named” which brings up a logistical question for warehouses.  For this example please assume we are following Path #1 and we have 3 SKUs (A, B, C) in a shipment and that they result in 6 boxes for each SKU (*with a few mixed), 18 boxes in total.  They look like this:

**EXAMPLE BOX CONFIGURATION:**

B1 = [  A  ]    B2 = [ A ]    B3 = [ A  ]    B4 = [  A  ]    B5 = [ A ]    B6 = [ A+B  ]

B7 = [  B  ]    B8 = [ B ]    B9 = [ B  ]    B10 = [ B  ]    B11 = [ B ]    B12 = [ B+C  ]

B13 = [  C  ]    B14 = [ C ]    B15 = [ C  ]    B16 = [  C  ]    B17 = [ C ]    B18 = [ C  ]

**LEGEND for above:**

B1 = [ A ] means that BOX 1 contains ITEM A.
B6 = [ A + B ] means that BOX 6 contains ITEM A+ B

Now let’s say we provide the boxes to Amazon through the “setPackingInformation” call and we provide the boxes and quantities in the exact order above.

NOTE during this step (setPackingInformation)
We cannot provide a box name or an internal box identifier to Amazon (functionality does not exist)
Amazon does not immediately return an internal Amazon box ID to us during this call that we can directly associate with a specific box above (** I do realize Amazon does attach a box ID but it is not available at this step)

Later in the process, Amazon splits our boxes into 3 shipments and sends out labels in a PDF file.   Those labels, I believe, have these identifiers on them assigned by Amazon internally.

**AMAZON BOX LABEL IDENTIFIERS PRINTED ON THE LABELS:**

P1 - B1  ,  P1 - B2  … P1 - B6
P2 - B1  ,   P2 - B2  … P1 - B6
P3 - B1  ,   P3 - B2  … P1 - B6

QUESTION 5a:  Does Amazon keep our box order provided through “setPackingInformation” when printing labels?  ** NOTE: This would only be partially useful due to center splits.   For example if we are sending to a single center, would the box labels come out in the exact order passed in setPackingInformation, or could it be “random”?

QUESTION 5b:  Please let us know how we accurately match up the Amazon box labels to our boxes?   Remember that we cannot provide an internal box identifier during setPackingInformation.   It is most likely that “P1 - B3” (ie: third box in Packing Group 1) is going to be box B7 – however we have no guarantee of that.    NOTE: even if box order is kept by Amazon (from the setPackingInformation process) this won’t help at this step if there are multiple shipments, and it also won’t help if box order accidentally changed (let’s say a warehouse employee had to move a box during the process).

We rely on matching by box contents (for example that a box contains “SKU A” or “SKU B”) because the Amazon provided shipping label only says “MIXED SKUS” for Boxes #6 and #12 (it does not list the full contents) so there would be no way to tell which label to put on those two boxes.   And even if Amazon did provide the full box contents, the font is so small and with a large # of mixed SKUs it would get very tedious.

Our operation regularly deals with mixed SKUs and shipments > 50 boxes, as well as multiple different “inbound plans” at one time.   This is why it is critical we need to provide an INTERNAL BOX IDENTIFIER to Amazon during setPackingInformation that is then printed on the resulting labels for matching ** OR ** Amazon returning a very specific Amazon identifier that we can somehow match to a specific box.   The only workaround would be to process only one single SKU in a shipment – however this is very inefficient as our shipping costs would skyrocket due to very small shipments (+ extra labor of shipping).   We maximize shipments > 500 lbs to a center so that we minimize shipping costs.

**Comment by lightningja** [comment]
Second this request. It will be really helpful that SP-API returns box Ids that can correspond to the original submitted boxes

**Reply by MetaMan2023** [reply]
@lightningja 

 I'm working trying to reach someone internally to discuss this issue further (I am an Amazon reseller).  If you can answer these it will be helpful in my discussions as I can have case studies from other sellers. 

* How are you (or do you) plan to handle the "labels matching" issue?

* How impactful will this be to your operation or your decision on how to proceed (ie: are you having to make compromises in shipping choices, or which "path" you take, etc) -- how much labor difference do you think this will take or human shipping errors that may occur?

* Which "path" are you leaning towards in your process (left path - items known up front, or right path, boxes packed later?)

* Do you use ship APC UPS or LTL?

Thanks

**Reply by MetaMan2023** [reply]
@lightningja 

FYI I have a thread to track everyone's status on the new API transition and issues -- check it out if you get a chance.

https://github.com/amzn/selling-partner-api-models/discussions/4427

**Reply by lightningja** [reply]
@MetaMan2023 thanks for the response.  To answer your questions: 
A rough implementation is to match boxes' weight/dimensions/items returned by sp-api with the original list of boxes. For large shipments this incurs more computation time as well as potential human errors. ideally sp-api could return boxes in the order of the original boxes submitted. That would make processing very fast as it is once scan rather than O(nm). 

I'm leaning towards left paths in most cases.  and we do both UPS and LTL

**Comment by MetaMan2023** [comment]
Thanks for replying on this LN.    In our situation it's basically a necessary requirement.

In a perfect scenario, we should be able to "name" boxes when we provide them in the API with some kind of internal (seller based) identifier -- for example when I tell Amazon a box like:

20 lbs
10" x 20" x 15"
Name: "SELLERBOX1"

and later in the process either:

(A) Amazon would print this box name on the actual FBA 4x6 label in large, clear writing (maybe on the bottom) so we can match it up ** PREFERRED **

(B) And/or they would at a minimum provide the ability for us to print a cross reference table -- this would be easy, when we ask for the box details back from Amazon they can return:

Box #1:
Amazon Box ID = XXXXX
Seller Box Name = YYYYY

And we can use that data to print a cross reference sheet or use it so that when we scan the internal seller box name on our boxes (internal barcode) we can then match up and request the appropriate label from Amazon to put on the box.

**Comment by puppsupr** [comment]
Hi,
Thank you for proposing an enhancement. This has been heard and will be improved in future, however, we do not have an estimated time of when the implementation will be complete.
As soon as this occurs, we will post an announcement in the [Announcements Section](https://developer-docs.amazon.com/sp-api/changelog) of our documentation site.
You can also review our [Release Notes](https://developer-docs.amazon.com/sp-api/page/release-notes) regarding all changes made to our APIs and subscribe to our [RSS feed](https://developer-docs.amazon.com/sp-api/changelog.rss) to get notified as soon as there is an update.

**Reply by MetaMan2023** [reply]
Thank you for the response and for adding it to the list.

Can I please request that before you implement it, you contact sellers for suggestions / real use cases?   I would be happy to assist with ideas.

I would hate for an upgrade to occur that does not solve the problem.

As a secondary note, this should be considered a highly important item as it almost makes the process unusable for larger scale sellers.

**Reply by JimMcKenzieSmith** [reply]
I 100% agree with everything @MetaMan2023 has said.  I work for a prep center that serves many large scale sellers.  We cannot function on v2024-03-20 without this feature.  **This is urgently important.**  On the left hand side of [the flow chart](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-use-case-guide#fulfillment-inbound-shipment-workflow), where "box packing information is known", there MUST be a mechanism to match up our internal box ID's to the resulting Amazon box labels.  We need to be able to match this up AFTER the split happens.

**Reply by MetaMan2023** [reply]
@JimMcKenzieSmith until just now, I thought I had figured out a tedious work-around -- I was going to make a cross-reference sheet using Amazon boxes (returned from ListInboundPlanBoxes) and my box IDs by matching up the quantities in the boxes.

I just realized that "ListInboundPlanBoxes" only returns the boxes for an INBOUND PLAN, not for a SHIPMENT **... if you have chosen to send items to more than one FBA center (ie: if your placement option has more than one center in it), I don't know if there is any way to match boxes up without some painstaking process of trying to match up box quantities with how many units Amazon wants to go to a center -- and I'm thinking that for a large # of boxes that could become computationally a nightmare.

** Does anyone know if you follow the left path and choose a 2 center split in Step 7, does ListInboundPlanBoxes "templateName" dynamically change to at least tell you which box goes to which center?   Also, if it does -- how do you determine which P# is for which shipment?

It seems clear not only do we need to name boxes, but there needs to be a "ListShipmentBoxes" call.

I have been at this for weeks, and my brain hurts trying to sort this out.

I have come to the conclusion that:

(A) If you want to use Amazon Partnered Carrier you must follow the left path.  If you are following the left hand path (where box content info is provided up front), the only realistic option is to send to a single center (and pay the full IPF).

(B) If you are following the right path (where you don't pack boxes up front), it may be slightly easier (though still a nightmare) but you cannot use Amazon Partnered Carrier.

@puppsupr

**Reply by JimMcKenzieSmith** [reply]
@MetaMan2023 I was digging deeper into the API, and came across [generateShipmentContentUpdatePreviews](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#post-inboundfba2024-03-20inboundplansinboundplanidshipmentsshipmentidcontentupdatepreviews). Have you looked at that endpoint?  It seems to eventually take a packageId (box ID) in a [BoxUpdateInput](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#boxupdateinput) array.  But, this endpoint is not documented very well in terms of how, why, and when it should be used. Thoughts?  Have you looked at it already?  Tested it?

**Reply by MetaMan2023** [reply]
@JimMcKenzieSmith I just ran a test to see what "listInboundPlanBoxes" does AFTER the shipments are confirmed. The templateName does NOT change.   (see details below).   Will look into your question now and reply in a few.

====

Following the left path (box info up front) I have created a shipment with the API going to two centers.   After the "confirmPlacementOption" the two shipments are created -- however the "listInboundPlanBoxes" function does NOT indicate which box goes with which shipment.  I had 5 boxes initially -- 3 going to one center and 2 to another.   Since ListInboundPlanBoxes only provides boxes at the inboundShipmentPlan level (not shipmentId level), the boxes are not "named" (ie: templateName) appropriately for the split, the all say "P1 - B#"... even after I confirm Placement Options which confirms the split.

**Reply by MetaMan2023** [reply]
> @MetaMan2023 I was digging deeper into the API, and came across [generateShipmentContentUpdatePreviews](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#post-inboundfba2024-03-20inboundplansinboundplanidshipmentsshipmentidcontentupdatepreviews). Have you looked at that endpoint? It seems to eventually take a packageId (box ID) in a [BoxUpdateInput](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#boxupdateinput) array. But, this endpoint is not documented very well in terms of how, why, and when it should be used. Thoughts? Have you looked at it already? Tested it?

I took a look at that function -- I have no idea what it does or what it is for... the documentation wording is awful.  Best I can tell it is for updating the box content at some later point in the process?   Also "packageId" seems to be generated by Amazon -- if you call "listInboundPlanBoxes" you will see the package I'd and it looks like this:

    "packageId" : "pk4444abcd-1234-abcd-5678-12adsfadsfa",

I was hoping to use packageId to match up labels in some way but again the issue is that even after creating the splits, there is no way to tell which labels goes with a box.

**Reply by JimMcKenzieSmith** [reply]
@MetaMan2023 Sounds good.  There's also a [listShipmentContentUpdatePreviews](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#get-inboundfba2024-03-20inboundplansinboundplanidshipmentsshipmentidcontentupdatepreviews), which seems similar to your earlier question about "ListShipmentBoxes".  Although it is a really strange name.  I haven't gotten quite far enough with the api to actually exercise that endpoint and get sample data.

**Reply by JimMcKenzieSmith** [reply]
> I was hoping to use packageId to match up labels in some way but again the issue is that even after creating the splits, there is no way to tell which labels goes with a box.

Ya, we're back to square one.  We need a seller box ID to be able to match up with.

**Reply by MetaMan2023** [reply]
> Hi, Thank you for proposing an enhancement. This has been heard and will be improved in future, however, we do not have an estimated time of when the implementation will be complete. As soon as this occurs, we will post an announcement in the [Announcements Section](https://developer-docs.amazon.com/sp-api/changelog) of our documentation site. You can also review our [Release Notes](https://developer-docs.amazon.com/sp-api/page/release-notes) regarding all changes made to our APIs and subscribe to our [RSS feed](https://developer-docs.amazon.com/sp-api/changelog.rss) to get notified as soon as there is an update.

@puppsupr 

Another request that goes along with this -- the function listInboundPlanBoxes needs to be updated as well with the following:

* listInboundPlanBoxes needs to return which SHIPMENT a box is assigned to -- for example after we choose to split to two centers in the left hand path and confirm, "listInboundPlanBoxes" should have an identifier to indicate which shipment the box goes with.  This is critical for creating pallets.   Right now there is no identifier to know which shipment Amazon has assigned boxes (though that information is clearly available).

* listInboundPlanBoxes needs to include the customer seller box identifier (discussed here).   So if we supply a custom box identifier in the setPackingInformation step then it should also appear in the listInboundPlanBoxes.

* Our custom box identifier should also appear in the CSV/PDF that becomes available for download in Seller Central

* Our custom box identifier should appear *LARGE* when listed on the FBA labels...  suggestion is that it is printed on the bottom of the label as there is plenty of room of a 4x6 (almost half the label is empty). 

* As a TEMPORARY fix, can you please make the "template name" that Amazon provides on the labels bigger?  Maybe large font and move that to the bottom (or repeat it) in the area you may ultimately print a seller custom box identifier?

* Please see my other GitHub thread on a bug with the CSV download -- the columns do not line up.   This spreadsheet download is the only information available that provides any insight into which box goes in which shipment.
*

**Reply by Cw733** [reply]
This functionality  has been missing even in the v0 API, but at least you could submit the CartonContentsFeed in a certain order and the shipping labels would generate in that order.

It's absolutely absurd that this functionality isn't present.  I submitted this as a problem over a year ago with the v0 API.

**Comment by JimMcKenzieSmith** [comment]
> Best I can tell it is for updating the box content at some later point in the process? Also "packageId" seems to be generated by Amazon

I was hoping that the packageId was something the seller could provide. But, given that you found that a package ID is Amazon generated, then this statement:

> Primary key to uniquely identify a Box Package. PackageId must be provided if the intent is to update an existing box. Adding a new box will not require providing this value. Any existing PackageIds not provided will be treated as to-be-removed

... seems to refer to the Amazon primary key.  So, not something we can use or match up with.

**Reply by MetaMan2023** [reply]
> > Best I can tell it is for updating the box content at some later point in the process? Also "packageId" seems to be generated by Amazon
> 
> I was hoping that the packageId was something the seller could provide. But, given that you found that a package ID is Amazon generated, then this statement:
> 
> > Primary key to uniquely identify a Box Package. PackageId must be provided if the intent is to update an existing box. Adding a new box will not require providing this value. Any existing PackageIds not provided will be treated as to-be-removed
> 
> ... seems to refer to the Amazon primary key. So, not something we can use or match up with.

The only use (currently) it may have is there is a function being carried over from v0 that you can print box labels and it may take a single box identifier -- at least that is how I read the functionality however I do not know if it works.

Are there any other forums for this stuff?  I am baffled out of millions of Amazon sellers that there are so few people talking about these issues.  The Amazon Seller forums are pretty much useless.  Here there are more programmers but again, very low volume.

We have paid SAS Support within Amazon and even with that, I get to pass questions back and forth through a third party to the API team and I get incomplete answers back days later... nobody can explain how to actually use this system in the real world.

**Reply by JimMcKenzieSmith** [reply]
@MetaMan2023 I found [this reddit thread](https://www.reddit.com/r/FulfillmentByAmazon/comments/1fkhwdu/has_anyone_switched_from_the_v0_fba_api_to_the/?rdt=45770).  And [this seller forum thread](https://sellercentral.amazon.com/seller-forums/discussions/t/5562b7df-eb2b-4531-8e6a-bfd5990cefa3?postId=95f54b99-6b9c-440f-8fef-3bca15141379) that I commented on 5 hours ago.  But, you are exactly right, very low volume.  Not much chatter for as many issues as this API has.  The December 20 deadline is ridiculous... they are going to terminate v0 on a Friday and then go on Christmas break... makes complete sense.  
I wonder if this issue should have it's own Amazon seller forum thread.  Just to make as much noise as possible in as many places as possible.

**Reply by MetaMan2023** [reply]
@JimMcKenzieSmith 

https://sellercentral.amazon.com/seller-forums/discussions/t/b3be8e49-9a5f-4f7c-a745-944f6c931875?postId=dec8cc01-200f-4c7f-95d8-2397c63697ee

Please see my post from a few weeks ago -- if you can reply and thumbs up it, may get some traction

**Comment by puppsupr** [comment]
While we improve this experience, the current workaround is to use [listInboundPlanBoxes](https://developer-docs.amazon.com/sp-api/docs/fulfillment-inbound-api-v2024-03-20-reference#listinboundplanboxes), then match the templateName with getLabels and box IDs.

**Reply by MetaMan2023** [reply]
@puppsupr 

Thank you for this information.

Can you please take a look at the getLabels (v0) function -- we are having an issue with the query called PackageLabelsToPrint

The function does not seem to care what we put in that field -- it gives no errors (even if we put incorrect data), nor does it seem to return any information based upon it.  We have tried both various different box Id's from the newer API as well.  No matter what we enter in that field, it returns the full list of labels.

Can you please see if this field is working or provide an example of what information to pass through it.

Also we are wondering -- if it does work as described in the documentation and we can provide a string of box id's to print... will those labels come out in the order we specify in that field?  (this would be important for warehouse efficiency as we would like to print the labels in a very specific order to correspond to how our boxes are laid out on our lines).

**Reply by MetaMan2023** [reply]
UPDATE for anyone interested -- get labels v0 will print a single label from the new API -- you need to specify:

**LabelType=UNIQUE**

and  PackageLabelsToPrint  must use the box label format Amazon provides which looks iike this:

[FBAXXXX]U[#####]

For example:

getLabels (v0) needs:

ShipmentId = [FBAXXXXX]
LabelType=UNIQUE
PackageLabelsToPrint = [FBAXXXXU00001],[FBAXXXXU00002],[FBAXXXXU00003]...

It will print those specific 3 labels from that FBA shipment -- however note, it will NOT current retain the order you specify.

Request to @puppsupr  the ability to retain the label order we request.

It's important when we have large shipments with hundreds of labels.   There are extremely tedious workarounds such as pulling labels one at a time then sorting them (in our own way) and re-combining them together but this will take forever making hundreds of API calls to Amazon, 2 for each label (one for the API call, one for the HTTP request).

**Comment by puppsupr** [comment]
HI @MetaMan2023,
Can you please take a look at the getLabels (v0) function -- we are having an issue with the query called PackageLabelsToPrint -> Are you saying that irrespective of the boxIDs it prints all the labels ? Ex, if you provide only 1 boxID, still is it printing all the labels ?

Regarding the order, I have looked into and it doesn't print in the order you specify. Its ascending/descending.

**Reply by MetaMan2023** [reply]
> HI @MetaMan2023, Can you please take a look at the getLabels (v0) function -- we are having an issue with the query called PackageLabelsToPrint -> Are you saying that irrespective of the boxIDs it prints all the labels ? Ex, if you provide only 1 boxID, still is it printing all the labels ?
> 
> Regarding the order, I have looked into and it doesn't print in the order you specify. Its ascending/descending.

We were able to get single labels to work after using your suggestion of "UNIQUE" label type.   As you noted, the labels print in an internal Amazon sorting format.   I've noted in other threads a request to allow for a custom sorting format.  This is very important for large quantities of labels as they get physically hard to manage -- every 100 labels = 50 boxes = 50 linear feet of labels to manage and having a custom sort order would allow us to peel the labels in order.

**Reply by puppsupr** [reply]
Your feedback has been passed on to the team.

**Comment by pitorian** [comment]
Does anyone have a work around for this, Amazon seems to be putting box contents into random boxes even when we set box 1 as FBAIDU00001 and they will place the contents into Box FBAIDU00005, its a complete mess this new API

This is our Box 9, The New Amazon API decided to place the contents of our Box 9 into FBAIDU00005 (i.e. which would be our Box 5)

How do we tell the New API that the contents are for Box 9, and to alloacate the FBAIDU00009 to it?
![FBAID-BOXID](https://github.com/user-attachments/assets/a5c57d97-6781-4af1-9fa9-9ecba9137370)

**Reply by MetaMan2023** [reply]
> Does anyone have a work around for this, Amazon seems to be putting box contents into random boxes even when we set box 1 as FBAIDU00001 and they will place the contents into Box FBAIDU00005, its a complete mess this new API
> 
> This is our Box 9, The New Amazon API decided to place the contents of our Box 9 into FBAIDU00005 (i.e. which would be our Box 5)
> 
> How do we tell the New API that the contents are for Box 9, and to alloacate the FBAIDU00009 to it? ![FBAID-BOXID](https://private-user-images.githubusercontent.com/34772476/406003085-a5c57d97-6781-4af1-9fa9-9ecba9137370.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc2MzYzMDIsIm5iZiI6MTczNzYzNjAwMiwicGF0aCI6Ii8zNDc3MjQ3Ni80MDYwMDMwODUtYTVjNTdkOTctNjc4MS00YWYxLTlmYTktOWVjYmE5MTM3MzcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIzVDEyNDAwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiNjdmMzg4NjIxNWQyNGI0Mzk1NWEyY2Y3MTM3YzliNTdkOTE0Mjc5OWM3NzViNTJkYjVhOThiNjVkOWNmOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.aJGQL54ZPXeAtKlyLFZI2zkxFFJZnDH7COcaBx5SmNc)

We have created a work-around for this issue, however it's very time consuming & complicated. 

**This is what we do:**

* Our packers pack boxes -- we keep track of all SKUs and quantities put into a particular box.  Each box has an internal Box ID of our own format.   We track in a database the box ID, SKU and QTY (and later the weights and dimensions of the box)

* We create the shipment plan and shipments through the API

* For each shipment, we request the inventory that Amazon has assigned to the boxes -- For each box, Amazon returns the SKU, QTY and Amazon's box IDs -- but as you are aware not OUR internal Box ID.

* Now we have two lists -- OUR list of boxes with OUR box IDs and Amazon's list of boxes with AMAZON's box IDs.   We match these up one by one in software -- basically we grab one of Amazon's boxes and search through our list for a matching box (ie: same QTY and SKU in the box -- making sure to take into account mixed SKUs as well).   Now we have a matching of Amazon's Box IDs to our Box IDs

* We use the information above to print a cross table -- basically a printable spreadsheet that says "OUR BOX ID = AMAZON BOX ID" -- we put some other information on it for reference like box weight and shipment ID.    (NOTE: You can get the FBA SHipment ID from the Amazon Box ID -- it's the part before the U -- Amazon labels the boxes like "FBAXXXXXU000000#" so we take the part before the U.

Now this technically worked, however there is a physical issue with the labels -- the labels do not come out in the proper order that you have sent to Amazon.   There is currently no way to have labels come out in a seller defined order, they come the way Amazon sorts them.  Imagine if you have 100 boxes, that's a giant pile of 200 labels -- and those labels are sorted not based on the order of your boxes but in alphabetical order on the box template name Amazon assigned.

This leads us to the second part of the solution -- label sorting.  This is what we do:

* We use the getLabels from v0 to pull labels from Amazon -- however again, this does not a low for custom order -- however you *CAN* pull one label at a time.

* We now pull each label (using a separate call) one at a time -- if we have 100 boxes, we make 100 calls then download 100 PDFs of labels -- that's 200 calls (which oddly goes surprisingly fast since it's V0 -- it's probably the fastest thing in the entire process).

* As we request each label, we request it using Amazon's box ID, but by this time we know OUR box ID so we keep track of which label is assigned to our INTERNAL box ID (in a sense we rename the PDF file from Amazon's box ID to our Box ID).

* NOW TO SORT:   As background we have a huge pack room with conveyor belts and piles of boxes.  The boxes for these shipments may be all over the place in the room in random order.   The shipping coordinator takes a scanner and scans the boxes IN ORDER on the line into our system -- we use this order to sort the PDFs into the proper order.

* Now we combine those PDFs back into a huge 200 label (100 boxes x 2 labels for each box) PDF and print it out.   (TECH NOTE: We use "ghostscript" (program "gs" in Linux) which is very helpful -- we use it to  get page counts and combine files together.

 We STILL have the issue of actually matching the labels up, since Amazon labels have their ID and we have our ID -- but now the labels are at least in order so that as we move through the room, the next label on the printed strip of 4x6 labels should be for a near by box.
 
 Remember that cross reference table from way up above?   A modification we made is that we print out a mini 10-item cross table every 10 labels.   So our labels look like this"

CROSS TABLE LABEL FOR 10 BOXES...
[BOX 1 labels
BOX 2 labels
BOX 3 labels
...
BOX 10 labels
CROSS TABLE #2 for next 10 boxes...
BOX 11 labels
BOX 12 labels
...

That's our "work-around"...   We have looked into actually modifying the individual PDF files with our box ID written over the label -- it's technically possible but we have limitations on PDF manipulation software we use right now and the resulting PDF's were coming out huge.  We need to find an alternative PDF modifying program to pull this off.

See how easy that is? ** sarcasm ** 

I was hesitant to post the details above as I am concerned they will be used as justification the API is fine and does not need to be updated but I believe that so many people are having this issue I want to help out.

**Reply by pitorian** [reply]
Thank you for the above, it’s just a disaster the way this new API had been rolled out, its full of faults, bad error handling, and now no control over saying what box id has what content, leading to more delays in dispatching pallets, I do hope someone on the amazon api dev team is monitoring the channel @puppsupr can you advise them that box contents needs a package id in the feed when we submit it to the API

**Reply by MetaMan2023** [reply]
> Thank you for the above, it’s just a disaster the way this new API had been rolled out, its full of faults, bad error handling, and now no control over saying what box id has what content, leading to more delays in dispatching pallets, I do hope someone on the amazon api dev team is monitoring the channel @puppsupr can you advise them that box contents needs a package id in the feed when we submit it to the API

They are aware and acknowledged sellers concern on that -- it's in theory "on the radar" but no road map for that has been provided for if/when it will occur. 

The slow speed is a huge hurdle as troubleshooting is painful.

**Reply by owen109** [reply]
Does anyone have any new solutions for this? It seems labels functionally print in a random order(since it doesnt assign keep our box ordering we submit to the api), so there’s no easy way to tie our box numbers (Box 1, Box 2, Box 3, etc.), which we submit in the call, to the actual labels. Is the only way to do this currently to pull all the box contents and create a mapping from one to the other? This seems very convoluted. @puppsupr, what would be the recommended solution for this? the goal being: submit a call to the inbound api with say 5 boxes and in the end when those labels are printed have them be in the same ordering they were submitted

**Reply by MetaMan2023** [reply]
Unfortunately the only way currently to do this is to:

Download a list of all boxes/contents/weights/dims from Amazon (meaning the list as they have it)

Match up each box with your own internal list -- marking off each one as you match it up.   You then will have a mapping of your internal box ID to their box ID.   Note you have to be careful here, for example if every single box is identical weight/dim/SKU, etc. you have to make sure to keep this mapping "forever" once you do it -- or write your code in a way that first sorts each list (yours and theirs) and checks off in a special order.  Otherwise you could have different mappings each time you do it.  (I hope that makes sense). 

For the labels, we literally pull each label PDF individually from Amazon, one by one, then match them up and recombine the PDF's together into one big PDF.  It's a royal pain in the *** to code but it does work.

If they allowed sorting labels and allowed us to provide a custom box ID for each box, it would have saved days of coding and simplified things.   It's "on the list of things to consider in the future" they say.

**Reply by owen109** [reply]
Appreciate the insights, a few things in your workflow i didn't think of so thats helpful! That list of "things to consider in the future" seems to be ever growing. It's very frustrating something as simple as printing labels in order requires all of this

**Comment by andremaytorena** [comment]
Hey everyone, I see this issue has been going on for months, a few people offered solutions like matching the dimensions/weight/sku of our internal boxes list to the amazon one to figure out the box ID's, but this wouldn't work if you'd like to use 2D BARCODES since you don't set the sky's and thus when you list the boxes from Amazon you wouldn't be able to see the sky's and matching based off dimensions and weight alone wouldn't be a good idea as that could present many errors with mismatching incorrectly. Has anyone found a workaround for this?

**Comment by Kalakata** [comment]
@puppsupr It’s been over a year, and this limitation still prevents a workflow that allows packing and labeling large quantities of boxes on the go. Can we expect this to change?

**Reply by Cw733** [reply]
I doubt it.  I have had this issue kicked up to dev (supposedly) for over a year by SAS Core (they are worthless), never a response.

The truth is that complaining about it to Amazon means nothing.  Either find a workaround or use a different workflow.  Sucks but after 20 years on Amazon that's the way it is, and it isn't going to change.

**Reply by Kalakata** [reply]
For the workflow I'm trying to create I believe there might be no workaround as the content, size and weight of the box is unknown until the moment the box is actually closed and a label should be printed and sticked to the box.

**Comment by SylvieManx** [comment]
I'm Manx, a third-party developer with many years of experience working within the Amazon ecosystem. I'm writing this email today to join you in speaking out against Amazon's newly introduced unreasonable pricing policy for its SP-API and to defend the survival of us small and medium-sized developers!

Since 2009, SP-API and its predecessor, MWS, have been freely available. This openness has fostered a thriving developer ecosystem. The ERP, inventory management, and data analytics tools we've developed have become indispensable operational support for millions of Amazon sellers. However, without consulting any developers, Amazon suddenly announced a fixed subscription fee of $1400 per year starting January 31, 2026, with an additional monthly fee based on GET requests starting April 30. This unilateral and mandatory charging completely disregards the legitimate rights of developers!

I. Three Core Injustices in the Pricing Policy, Targeting the Lifeline of Small and Medium-Sized Developers

1. The High Fixed Annual Fee: The $1400 annual subscription fee applies regardless of size or profitability. Even small and medium-sized developers serving only a few sellers and barely breaking even must bear this enormous fixed cost. Many features have very low profit margins, and this fee puts us in a desperate situation of "no service, no pay" or "losses."

2. Severe Disconnect Between Pricing and Service Quality: Amazon's API service has long suffered from stability issues. Many colleagues have encountered problems such as official demo errors, multiple failed permission requests, and call anomalies. Even more frustrating is that when we open a case for consultation, we often don't receive a valid response for one or two weeks. This combination of "low-quality service + high fees" completely violates the principles of fair business practices.

3. Forced Bundling Under a Monopolistic Position: SP-API is the only legal channel for connecting to Amazon's platform data; we have no alternatives. Amazon, leveraging its market dominance, forces all developers reliant on its ecosystem to accept unreasonable fees, essentially engaging in monopolistic practices that exclude market competition and squeeze developers' profits.

II. Collective action is the only way out: File a joint complaint with the FTC, demanding Amazon rectify its practices. Individual voices are easily ignored, but our combined strength is enough to drive change! The U.S. Federal Trade Commission (FTC) specifically tracks unfair business practices, and its Consumer Sentinel Network aggregates and analyzes all complaints to form a basis for regulatory action. Here, I urge every affected developer to immediately submit a complaint through the following official channels:

• Complaint Email: oig@ftc.gov

• Online Complaint Form (most convenient): reportfraud.ftc.gov

• Mailing Address: Federal Trade Commission, Consumer Sentinel Network, Washington, D.C. 20580

• Complaint Hotline: 1-877-438-4338 (English service) When filing a complaint, please emphasize the following core issues: Amazon's abuse of its SP-API monopoly to suddenly introduce unreasonable fees; the $1400 fixed annual fee squeezing the survival space of small and medium-sized developers; a serious mismatch between service quality and fees; and the lack of prior consultation and a transition period.

III. Our Demands: Reject Monopolistic Exploitation, Demand a Reasonable Fee Mechanism
We are not against reasonable fees, but rather we resist a "one-size-fits-all" monopolistic fee model. We strongly urge Amazon to:

1. Suspend the current pricing policy and publicly negotiate a new pricing plan with the global developer community;

2. Introduce a tiered pricing mechanism, offering low-threshold entry-level packages or fee reductions for small and medium-sized developers;

3. Improve service quality, ensure API stability, and clarify technical support response times (e.g., responding to cases within 24 hours);

4. Publicly disclose the basis for cost accounting, making pricing transparent and reasonable.

Dear colleagues, the prosperity of Amazon's ecosystem depends on the contributions and innovation of every developer. This current "discarding the millstone after it has served its purpose" pricing policy will not only force many small and medium-sized developers out of the market, but will also ultimately lead to fewer sellers, increased operating costs, and disrupt the balance of the entire ecosystem.

Please take immediate action and forward this email to more developer partners to jointly file a complaint with the FTC! Every complaint is a contribution; only by uniting can we make Amazon take our demands seriously and safeguard our legitimate rights!

We look forward to witnessing the arrival of fairness and justice together! Initiator: Manx
Contact Email: lyxdirect@163.com
